lib/WUI/http/upload_state.cpp

```diff
@@ -0,0 +1,649 @@
+     * Starts looking for specific substring.
+     *
+     * * Reset the buffer to be "empty".
+     * * Start looking for the given string.
+     * * Sensitive means if it should be case sensitive or not.
+     */
+    void start(const char *looking_for, bool sensitive) {
+        const size_t len = strlen(looking_for);
+        assert(len <= buffer.size());
+        this->looking_for = string_view(looking_for, len);
+        this->sensitive = sensitive;
+        this->len = 0;
+    }
+    /*
+     * Puts another character into the buffer.
+     *
+     * This potentially pushes the oldest one out in case the buffer is full.
+     *
+     * Returns if the substring has been found at this position or not. It may
+     * return Overflow, which means the string already fed in here doesn't
+     * match from the beginning. Depending on if you're looking for an
+     * exact/full match or substring anywhere, you may take this as a "No
+     * match" or "Keep looking".
+     */
+    Lookup feed(char c) {
+        memmove(buffer.begin(), buffer.begin() + 1, buffer.size() - 1);
+        if (!sensitive) {
+            c = tolower(c);
+        }
+        buffer[buffer.size() - 1] = c;
+        const bool was_full = (len == buffer.size());
+        if (!was_full) {
+            len += 1;
+        }
+
+        const size_t suffix_len = std::min(static_cast<size_t>(len), looking_for.size());
+        string_view suffix(buffer.end() - suffix_len, suffix_len);
+
+        if (suffix == looking_for) {
+            return Lookup::Found;
+        } else if (was_full) {
+            return Lookup::Overflow;
+        } else {
+            return Lookup::NotYet;
+        }
+    }
+    // Is the buffer empty?
+    bool empty() const {
+        return len == 0;
+    }
+    /*
+     * Returns the part of buffer before the looking-for string (as much of it as fits).
+     *
+     * If the looking-for string is considered a separator, the thing before it
+     * is left in the buffer and can be accessed with this.
+     *
+     * The returned prefix can be shorter than whatever was fed into it before
+     * the looking-for separator (in case oldest chars no longer fit). This is
+     * OK in our use, as the buffer - separator is longer (due to other
+     * reasons) than the constant strings we look for before the separator, so
+     * in case something fell off the left side of the buffer, it would no
+     * longer match anyway.
+     */
+    string_view prefix() const {
+        assert(len >= looking_for.size());
+        return string_view(buffer.end() - len, len - looking_for.size());
+    }
+    /*
+     * Access the internal buffer.
+     *
+     * This can be used to recycle the buffer for different purposes (to save
+     * memory) while not being in used. The buffer or the functionality of
+     * looking for something are mutually exclusive (use one or the other, not
+     * both at the same time).
+     */
+    array<char, MAX_SUBTOKEN> &borrow_buf() {
```

Comment:
reusing memory is a good principle, but I'd recommend to secure the access to the memory area with some simple locking mechanism, which would assert in debug - so that the accidental usage of this buffer from multiple spots at once gets detected at least at runtime in debug mode if not at compile time.

Reply:
I've added something, but please check if you like it. I'm not entirely sure about it, since:

* The scope in which this class is used is local and rather small. So it's possible to check manually. Furthermore, messing it up would very likely break tests.
* The current solution adds some complexity and consumes more memory (not sure how much, alignment issues are in play).

In case one would get it wrong, it would throw an exception. I assume this would terminate the program with something in our case?

I was also thinking about a bool tracking „Am I borrowed?“ and returning some kind of RAII object to unlock it on drop. But then the caller would need to _store that RAII object_, which is problematic ‒ currently, it doesn't store the borrow, but pointers into it.

This is the place where I miss Rust's enums (that are somewhat close to the std::variant thing, but they have language syntax support to make them comfortable and can store the index of currently active variant in „niches“ ‒ bits that are unused in the actual stored values; besides the order of fields is unspecified, unlike C/C++, which gives the compiler the freedom to optimize by reordering them ‒ the thing I'm hoping to give the compiler by hiding the implementation comes for free there).

Reply:
:thinking: the variant solution looks too heavy. Morover, we have exceptions disabled, so who knows what it does at runtime. But the principle is there - having some kind of handle which checks if the count of simultaneous requests for the handle is <= 1.